### PR TITLE
hardening/2009_check_of_notification_http_url_in_post_v2_subscriptions

### PR DESCRIPTION
**Pull request by kzangeli**
Fixed issue #2009, for IPv4 only.
FIXME P4 added about doing the same check for IPv6 addresses (I am not 100% sure about this distinction between IPv4 and 6 ... I imagine the hostname 'myhost' could be used for an IPv6 address as well as an IPv4 address ... ? So, ... let's discuss/investigate this a little!)

### CODE REVIEW

File: src/lib/common/string.cpp [hunk 1]
```diff
@@ -213,6 +214,63 @@ int stringSplit(const std::string& in, char delimiter, std::vector<std::string>&
...
+*
+* hostnameIsValid - check a hostname for validity
+*
+* See https://en.wikipedia.org/wiki/Hostname
+*/
+static bool hostnameIsValid(const char* hostname, bool isIPv6)
+{
+  if (isIPv6 == true)  // No checks for IPv6
+  {
+    // FIXME P4: Implement hostnameIsValid for IPv6 addresses
```
Comment (fgalan) [inline src/lib/common/string.cpp]: Hostnames are the same for IPv4 or IPv6. It is the name resolver (e.g. `/etc/hosts`, the DNS subsystem) the one which will provide either a IPv4 or IPv6 (or even both... I don't remember right now the details).

Thus, is this is an actual hostname checking function, I think whe whole IPV6 hasn't nothing to do.

File: src/lib/common/string.cpp [hunk 2]
```diff
@@ -345,6 +405,21 @@ bool parseUrl(const std::string& url, std::string& host, int& port, std::string&
     }
   }
 
+  //
+  // Is 'port' a valid number?
+  // Remember the port is really an 'unsigned short', so numbers over MAXSHORT (0xFFFF) are
+  // not valid. Negative port numbers cannot exist and 0 is reserved.
+  //
+  if ((port > MAXSHORT) || (port <= 0))
```
Comment (fgalan) [inline src/lib/common/string.cpp]: Valid port are a range betweeh 0 and 65365 (by desing of the TCP/IP stack). In our case it happens that MAXSHORT matches 65365 but it is an "implementation side effect". I think that using MAX_PORT (defined in limits.h) would be more explicit.

Comment (kzangeli) [inline src/lib/common/string.cpp]: Ah, if MAX_PORT exists (probably defined to MAXSHORT :-)) then I will use it, of course.

Zero on the other hand I prefer not to let pass, as it seems to be a reserved value and cannot be used.
See https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers

[ Also, conveniently enough: atoi("invalid string") == 0 ]

File: src/lib/common/string.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/common/string.cpp]: ok
I will trust you on your word, I myself am clueless :-)

File: src/lib/common/string.cpp (hunk 2 above)
Comment (fgalan) [inline src/lib/common/string.cpp]: Zero is reserved. Fine.

If MAX_PORT exists then use it, but if doesn't exist, please define it and use it ;)

Comment (kzangeli) [inline src/lib/common/string.cpp]: Fixed in 900322f

File: src/lib/common/string.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/common/string.cpp]: Fixed in a8e4057

**Comment by fgalan** [comment]
LGTM

**Pull request merged**